src/nbl/video/ILogicalDevice.cpp

```diff
@@ -1050,14 +1094,47 @@ bool ILogicalDevice::createRayTracingPipelines(IGPUPipelineCache* const pipeline
             NBL_LOG_ERROR("Invalid maxRecursionDepth. maxRecursionDepth(%u) exceed the limits(%u)", param.cached.maxRecursionDepth, limits.maxRayRecursionDepth);
             return false;
         }
-        if (param.getShaders().empty())
+
+        SpirvDebloatTask debloatTask(m_spirvDebloater.get(), m_logger);
+        debloatTask.insertEntryPoint(param.shaderGroups.raygen, hlsl::ShaderStage::ESS_RAYGEN);
+        for (const auto& miss : param.shaderGroups.misses)
+            debloatTask.insertEntryPoint(miss, hlsl::ShaderStage::ESS_MISS);
+        for (const auto& hit : param.shaderGroups.hits)
         {
-            NBL_LOG_ERROR("Pipeline must have at least one shader.");
-            return false;
+            debloatTask.insertEntryPoint(hit.closestHit, hlsl::ShaderStage::ESS_CLOSEST_HIT);
+            debloatTask.insertEntryPoint(hit.anyHit, hlsl::ShaderStage::ESS_ANY_HIT);
+            debloatTask.insertEntryPoint(hit.intersection, hlsl::ShaderStage::ESS_INTERSECTION);
+        }
+        for (const auto& callable : param.shaderGroups.callables)
+            debloatTask.insertEntryPoint(callable, hlsl::ShaderStage::ESS_CALLABLE);
+
+        newParams[ix] = param;
+        newParams[ix].shaderGroups.raygen = debloatTask.debloat(param.shaderGroups.raygen, debloatedShaders);
+
+        newParams[ix].shaderGroups.misses = { debloatedMissSpecData, param.shaderGroups.misses.size() };
+        for (const auto& miss: param.shaderGroups.misses)
+        {
+            *debloatedMissSpecData = debloatTask.debloat(miss, debloatedShaders);
+            debloatedMissSpecData++;
         }
 
-        newParams[ix].shaders = std::span(outShaderSpecs, param.getShaders().size());
-        debloatShaders(*m_spirvDebloater.get(), param.getShaders(), debloatedShaders, outShaderSpecs, m_logger);
+        newParams[ix].shaderGroups.hits = { debloatedHitSpecData, param.shaderGroups.hits.size() };
+        for (const auto& hit: param.shaderGroups.hits)
+        {
+            *debloatedHitSpecData = {
+                .closestHit = debloatTask.debloat(hit.closestHit, debloatedShaders),
+                .intersection = debloatTask.debloat(hit.intersection, debloatedShaders),
+                .anyHit = debloatTask.debloat(hit.anyHit, debloatedShaders),
+            };
+            debloatedHitSpecData++;
+        }
+
+        newParams[ix].shaderGroups.callables = { debloatedCallableSpecData, param.shaderGroups.callables.size() };
```

Comment:
btw span has an implicit constructor from vector, you can just write `span = vector;` instead of `span = {vector.data(),vector.size()}`